cli/flox-watchdog/src/process.rs

```diff
@@ -0,0 +1,414 @@
+    pub fn stop_process(mut child: Child) {
+        child.kill().expect("failed to kill");
+        child.wait().expect("failed to wait");
+    }
+
+    /// Makes two Arc<AtomicBool>s to mimic the shutdown flags used by
+    /// the watchdog
+    pub fn shutdown_flags() -> (Arc<AtomicBool>, Arc<AtomicBool>) {
+        (
+            Arc::new(AtomicBool::new(false)),
+            Arc::new(AtomicBool::new(false)),
+        )
+    }
+
+    /// Writes a registry to a temporary file, adding an entry for the provided
+    /// path hash
+    pub fn path_for_registry_with_entry(path_hash: impl AsRef<str>) -> NamedTempFile {
+        let path = NamedTempFile::new().unwrap();
+        let mut reg = EnvRegistry::default();
+        let entry = RegistryEntry {
+            path_hash: String::from(path_hash.as_ref()),
+            path: PathBuf::from("foo"),
+            envs: vec![],
+            activations: HashSet::new(),
+        };
+        reg.entries.push(entry);
+        let string = serde_json::to_string(&reg).unwrap();
+        std::fs::write(&path, string).unwrap();
+        path
+    }
+
+    /// Wait some attempts for the process to reach the desired state
+    fn poll_until_state(state: ProcStatus, pid: ActivationPid) {
+        for _ in 0..10 {
+            if PidWatcher::read_pid_status(pid) == state {
+                return;
+            }
+            std::thread::sleep(Duration::from_millis(10));
+        }
+        panic!("never entered zombie state");
+    }
+
+    #[test]
+    fn reports_that_pid1_is_running() {
+        assert!(PidWatcher::pid_is_running(1.into()));
+    }
+
+    #[test]
+    fn detects_running_or_not_running_process() {
+        let proc = start_process();
+        let pid = proc.id() as i32;
+        assert!(PidWatcher::pid_is_running(pid.into()));
+        stop_process(proc);
+        assert!(!PidWatcher::pid_is_running(pid.into()));
+    }
+
+    #[test]
+    fn detects_zombie() {
+        let mut proc = Command::new("true").spawn().unwrap();
+        let pid = proc.id() as i32;
+        poll_until_state(ProcStatus::Zombie, pid.into());
+        assert!(!PidWatcher::pid_is_running(pid.into()));
+        assert_eq!(PidWatcher::read_pid_status(pid.into()), ProcStatus::Zombie);
+        proc.wait().unwrap();
+    }
+
+    #[test]
+    fn terminates_when_all_pids_terminate() {
+        let proc1 = start_process();
+        let pid1 = ActivationPid::from(proc1.id() as i32);
+        let proc2 = start_process();
+        let (terminate_flag, cleanup_flag) = shutdown_flags();
+        let path_hash = "abc";
+        let reg_path = path_for_registry_with_entry(&path_hash);
+        register_activation(&reg_path, &path_hash, pid1).unwrap();
+        let mut watcher =
+            PidWatcher::new(pid1, &reg_path, &path_hash, terminate_flag, cleanup_flag);
+        let wait_result = std::thread::scope(move |s| {
+            let procs_handle = s.spawn(|| {
+                std::thread::sleep(Duration::from_millis(100));
```

Comment:
The ordering is necessary to prevent the processes from being killed before the watchdog has started watching them. We can use a condvar/thread barrier instead.

Reply:
We do need to handle that case though right? I guess that would be a separate test. Nonblocking just checking my understanding